3rdparty/openvx/include/openvx_hal.hpp

```diff
@@ -151,30 +174,92 @@ struct vxImage
         addr.step_x = 1;
         addr.step_y = 1;
         void *ptrs[] = { (void*)data };
-        img = vxCreateImageFromHandle(ctx.ctx, VX_Traits<T>::Type, &addr, ptrs, VX_MEMORY_TYPE_HOST);
+        img = vxCreateImageFromHandle(ctx.ctx, VX_Traits<T>::ImgType, &addr, ptrs, VX_MEMORY_TYPE_HOST);
         vxErr::check(img);
     }
     ~vxImage()
     {
-        vxSwapImageHandle(img, NULL, NULL, 1);
+        vxErr::check(vxSwapImageHandle(img, NULL, NULL, 1));
         vxReleaseImage(&img);
     }
 };
 
+struct vxMatrix
+{
+    vx_matrix mtx;
+
+    template <typename T>
+    vxMatrix(vxContext &ctx, const T *data, int w, int h)
+    {
+        mtx = vxCreateMatrix(ctx.ctx, VX_Traits<T>::DataType, w, h);
+        vxErr::check(mtx);
+        vxErr::check(vxCopyMatrix(mtx, data, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST);
+    }
+    ~vxMatrix()
+    {
+        vxReleaseMatrix(&mtx);
+    }
+};
+
+struct vxConvolution
+{
+    vx_convolution cnv;
+
+    vxConvolution(vxContext &ctx, const short *data, int w, int h)
+    {
+        cnv = vxCreateConvolution(ctx.ctx, w, h);
+        vxErr::check(cnv);
+        vxErr::check(vxCopyConvolutionCoefficients(cnv, (void*)data, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST);
+    }
+    ~vxConvolution()
+    {
+        vxReleaseConvolution(&cnv);
+    }
+};
+
 //==================================================================================================
 // real code starts here
 // ...
 
-template <typename T>
-inline int ovx_hal_add(const T *a, size_t astep, const T *b, size_t bstep, T *c, size_t cstep, int w, int h)
+#define OVX_BINARY_OP(hal_func, ovx_call, ...)                                                                                   \
+template <typename T>                                                                                                            \
+inline int ovx_hal_##hal_func(const T *a, size_t astep, const T *b, size_t bstep, T *c, size_t cstep, int w, int h, __VA_ARGS__) \
```

Comment:
GCC does not like remove preceding comma in case of empty `__VA_ARGS__`: https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html
It should be `##__VA_ARGS__`, but I don't know whether VS support such syntax.